packages/api/src/controllers/project.ts

```diff
@@ -0,0 +1,157 @@
+import { Request, RequestHandler, Router } from "express";
+import { authorizer, validatePost } from "../middleware";
+import { db } from "../store";
+import { v4 as uuid } from "uuid";
+import {
+  makeNextHREF,
+  parseFilters,
+  parseOrder,
+  getS3PresignedUrl,
+  toStringValues,
+  pathJoin,
+  getObjectStoreS3Config,
+  reqUseReplica,
+  isValidBase64,
+  mapInputCreatorId,
+} from "./helpers";
+import sql from "sql-template-strings";
+import {
+  ForbiddenError,
+  UnprocessableEntityError,
+  NotFoundError,
+  BadRequestError,
+  InternalServerError,
+  UnauthorizedError,
+  NotImplementedError,
+} from "../store/errors";
+
+const app = Router();
+
+export async function getProject(req) {
+  const projectId = req.params.projectId || req.body.projectId;
+
+  const project = await db.project.get(projectId);
+  if (!project || project.deleted) {
+    throw new NotFoundError(`project not found`);
+  }
+
+  if (!req.user.admin && req.user.id !== project.userId) {
+    throw new ForbiddenError(`invalid user`);
+  }
+
+  return project;
+}
+
+const fieldsMap = {
+  id: `project.ID`,
+  name: { val: `project.data->>'name'`, type: "full-text" },
+  createdAt: { val: `project.data->'createdAt'`, type: "int" },
+  userId: `project.data->>'userId'`,
+} as const;
+
+app.get("/", authorizer({}), async (req, res) => {
+  let { limit, cursor, order, all, filters, count } = toStringValues(req.query);
+
+  if (isNaN(parseInt(limit))) {
+    limit = undefined;
+  }
+  if (!order) {
+    order = "updatedAt-true,createdAt-true";
+  }
+
+  const query = [...parseFilters(fieldsMap, filters)];
+
+  if (!req.user.admin || !all || all === "false") {
+    query.push(sql`project.data->>'deleted' IS NULL`);
+  }
+
+  let output: WithID<Project>[];
+  let newCursor: string;
+  if (req.user.admin) {
+    let fields =
+      " project.id as id, project.data as data, users.id as usersId, users.data as usersdata";
+    if (count) {
+      fields = fields + ", count(*) OVER() AS count";
+    }
+    const from = `project left join users on project.data->>'userId' = users.id`;
```

Comment:
Do you think it would be possible to put the SQL queries inside the table? I know it's the same in other controllers, but I wonder if it would make sense to abstract it and put inside the `store/*-table.ts`?

Reply:
We could do that - I just followed the existing patterns in other files.

Reply:
Ok, that's probably fine, we can try to clean it up for all controllers in some other PR.